spikeextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -349,46 +349,43 @@ def add_electrode_groups(recording: se.RecordingExtractor, nwbfile=None,
-                                                                         grp.get('device_name', default_dev_name)}}}
-                                    se.NwbRecordingExtractor.add_devices(recording, nwbfile, metadata=new_device)
-                                    print("Warning: device name not detected in attempted link to electrode group! "
-                                          "Automatically generating.")
-
-                                electrode_group_kwargs = dict(defaults, **grp)
-                                electrode_group_kwargs.pop('device_name')
-                                electrode_group_kwargs.update(
-                                    {'device': nwbfile.devices[grp.get('device_name', default_dev_name)]})
-                                nwbfile.create_electrode_group(**electrode_group_kwargs)
-                        else:
-                            print(f"Warning: Expected metadata['Ecephys']['ElectrodeGroup'][{j}] to be"
-                                  " a dictionary with keys 'name', 'description', 'location', and 'device'!"
-                                  f"Electrode Group [{j}] will not be created.")
-                else:
-                    print("Warning: metadata must be a list of dictionaries of the form"
-                          " metadata['Ecephys']['ElectrodeGroup'] = [{'name': my_name,"
-                          " 'description': my_description, 'location': electrode_location,"
-                          " 'device_name': my_device_name}, ...]!")
+        assert all([isinstance(x, dict) for x in metadata['Ecephys']['ElectrodeGroup']]), \
+            "Expected metadata['Ecephys']['ElectrodeGroup'] to be a list of dictionaries!"
+
+        for grp in metadata['Ecephys']['ElectrodeGroup']:
+            device_name = grp.get('device_name', defaults['device_name'])
+            if grp.get('name', defaults['name']) not in nwbfile.electrode_groups:
+                if device_name not in nwbfile.devices:
+                    new_device = dict(
+                                    Ecephys=dict(
+                                        Device=dict(
+                                            name=device_name
+                                        )
+                                    )
+                                 )
+                    se.NwbRecordingExtractor.add_devices(recording, nwbfile, metadata=new_device)
+                    print("Warning: device name not detected in attempted link to electrode group! "
+                          "Automatically generating.")
+                electrode_group_kwargs = dict(defaults, **grp)
+                electrode_group_kwargs.pop('device_name')
+                electrode_group_kwargs.update(dict(device=nwbfile.devices[device_name]))
+                nwbfile.create_electrode_group(**electrode_group_kwargs)
 
         if not nwbfile.electrode_groups:
             if len(nwbfile.devices.keys()) == 1:
```

Comment:
```suggestion
            if len(nwbfile.devices) == 1:
```

Reply:
Ah, right you are. The `len(dict(....))` returns the # of key-value pairs, so the keys themselves are not necessary for those types of checks. Fixed now